lib/defaultingress/api_ingress.go

```diff
@@ -0,0 +1,271 @@
+/*
+  This file is part of Astarte.
+
+  Copyright 2021 Ispirata Srl
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+package defaultingress
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	v1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/go-logr/logr"
+	"github.com/openlyinc/pointy"
+
+	commontypes "github.com/astarte-platform/astarte-kubernetes-operator/apis/api/commontypes"
+	apiv1alpha1 "github.com/astarte-platform/astarte-kubernetes-operator/apis/api/v1alpha1"
+	ingressv1alpha1 "github.com/astarte-platform/astarte-kubernetes-operator/apis/ingress/v1alpha1"
+	"github.com/astarte-platform/astarte-kubernetes-operator/lib/misc"
+)
+
+func EnsureAPIIngress(cr *ingressv1alpha1.AstarteDefaultIngress, parent *apiv1alpha1.Astarte, c client.Client, scheme *runtime.Scheme, log logr.Logger) error {
+	ingressName := getAPIIngressName(cr)
+	if !pointy.BoolValue(cr.Spec.API.Deploy, true) {
+		// We're not deploying the Ingress, so we're stopping here.
+		// However, maybe we have an Ingress to clean up?
+		ingress := &ingressv1alpha1.AstarteDefaultIngress{}
```

Comment:
I think this is conceptually wrong - instead of removing the deployments created by the ingress, you're triggering the deletion of the AstarteDefaultIngress object, basically committing suicide and wiping the CR. Instead, you should check whether a `networkingv1.Ingress` owned by the operator and matching the parameters exist and, in case, delete it.

Reply:
Thanks for spotting this! Fixed now